signer/core/signed_data.go

```diff
@@ -0,0 +1,714 @@
+	if len(sig) != 65 {
+		return common.Address{}, fmt.Errorf("signature must be 65 bytes long")
+	}
+	if sig[64] != 27 && sig[64] != 28 {
+		return common.Address{}, fmt.Errorf("invalid Ethereum signature (V is not 27 or 28)")
+	}
+	sig[64] -= 27 // Transform yellow paper V from 27/28 to 0/1
+	hash, _ := SignTextPlain(data)
+	rpk, err := crypto.SigToPub(hash, sig)
+	if err != nil {
+		return common.Address{}, err
+	}
+	return crypto.PubkeyToAddress(*rpk), nil
+}
+
+// UnmarshalValidatorData converts the bytes input to typed data
+func UnmarshalValidatorData(data interface{}) (ValidatorData, error) {
+	raw := data.(map[string]interface{})
+
+	addr, ok := raw["address"].(string)
+	addrBytes, err := hexutil.Decode(addr)
+	if err != nil {
+		return ValidatorData{}, err
+	}
+	if !ok || len(addrBytes) == 0 {
+		return ValidatorData{}, errors.New("validator address is undefined")
+	}
+
+	message, ok := raw["message"].(string)
+	messageBytes, err := hexutil.Decode(message)
+	if err != nil {
+		return ValidatorData{}, err
+	}
+	if !ok || len(messageBytes) == 0 {
+		return ValidatorData{}, errors.New("message is undefined")
+	}
+
+	return ValidatorData{
+		Address: common.BytesToAddress(addrBytes),
+		Message: messageBytes,
+	}, nil
+}
+
+// Validate make sure the types are sound
+func (typedData *TypedData) Validate() error {
+	if err := typedData.Types.Validate(); err != nil {
+		return err
+	}
+	if err := typedData.Domain.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Map generates a map version of the typed data
+func (typedData *TypedData) Map() map[string]interface{} {
+	dataMap := map[string]interface{}{
+		"types":       typedData.Types,
+		"domain":      typedData.Domain.Map(),
+		"primaryType": typedData.PrimaryType,
+		"message":     typedData.Message,
+	}
+	return dataMap
+}
+
+// Validate checks if the types object is conformant to the specs
+func (types *EIP712Types) Validate() error {
+	for typeKey, typeArr := range *types {
+		for _, typeObj := range typeArr {
+			typeVal := typeObj["type"]
+			if typeKey == typeVal {
+				return fmt.Errorf("type '%s' cannot reference itself", typeVal)
+			}
+			firstChar := []rune(typeVal)[0]
+			if unicode.IsUpper(firstChar) {
+				if (*types)[typeVal] == nil {
+					return fmt.Errorf("referenced type '%s' is undefined", typeVal)
+				}
+			} else {
+				if !isStandardTypeStr(typeVal) {
+					if (*types)[typeVal] != nil {
+						return fmt.Errorf("referenced type '%s' must be capitalized", typeVal)
+					} else {
+						return fmt.Errorf("unknown atomic type '%s'", typeVal)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// isStandardType checks if the given type is a EIP712 conformant type
+func isStandardTypeStr(encType string) bool {
+	// Atomic types
+	exp, _ := regexp.Compile(`^(address|bool|bytes|string)$`)
+	if exp.MatchString(encType) {
+		return true
+	}
+
+	// Dynamic types
+	exp, _ = regexp.Compile(`^(bytes|int|uint)(\d+)$`)
```

Comment:
And here's the typical problem with regexps. It's really tricky to get right. So this would e.g. allow `uint999999999999999999`